client/driver/logging/rotator.go

```diff
@@ -0,0 +1,218 @@
+package logging
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+)
+
+var (
+	bufSize = 32
+	buf     = make([]byte, bufSize)
+)
+
+// FileRotator writes bytes to a rotated set of files
+type FileRotator struct {
+	// MaxFiles is the maximum number of rotated files allowed in a path
+	MaxFiles int
+	// FileSize is the size a rotated file is allowed to grow
+	FileSize int64
+
+	// path is the path on the file system where the rotated set of files are
+	// opened
+	path string
+	// baseFileName is the base file name of the rotated files
+	baseFileName string
+
+	// logFileIdx is the current index of the rotated files
+	logFileIdx int
+	// oldestLogFileIdx is the index of the oldest log file in a path
+	oldestLogFileIdx int
+
+	// currentFile is the file that is currently getting written
+	currentFile *os.File
+	// currentWr is the number of bytes written to the current file
+	currentWr int64
+
+	logger  *log.Logger
+	purgeCh chan struct{}
+}
+
+// NewFileRotator returns a new file rotator
+func NewFileRotator(path string, baseFile string, maxFiles int,
+	fileSize int64, logger *log.Logger) (*FileRotator, error) {
+	rotator := &FileRotator{
+		MaxFiles: maxFiles,
+		FileSize: fileSize,
+
+		path:         path,
+		baseFileName: baseFile,
+
+		logger:  logger,
+		purgeCh: make(chan struct{}, 1),
+	}
+	if err := rotator.lastFile(); err != nil {
+		return nil, err
+	}
+	go rotator.purgeOldFiles()
+	return rotator, nil
+}
+
+// Write writes a byte array to a file and rotates the file if it's size becomes
+// equal to the maximum size the user has defined.
+func (f *FileRotator) Write(p []byte) (n int, err error) {
+	n = 0
+	var nw int
+
+	for n < len(p) {
+		// check if we still have space in the current file, otherwise close and
+		// open the next file
+		if f.currentWr >= f.FileSize {
+			f.currentFile.Close()
+			if err := f.nextFile(); err != nil {
+				return 0, err
+			}
+		}
+		// calculate the remaining size on this file
+		remainingSize := f.FileSize - f.currentWr
+		// check if the number of bytes that we have to write is less than the
+		// remaining size of the file
+		if remainingSize < int64(len(p[n:])) {
+			// write the number of bytes that we can write on the current file
+			li := int64(n) + remainingSize
+			nw, err = f.currentFile.Write(p[n:li])
+		} else {
+			// write all the bytes in the current file
+			nw, err = f.currentFile.Write(p[n:])
+		}
+
+		// increment the number of bytes written to the current file in this
```

Comment:
`n` is the total written, not "to the current file". I think you are trying to say the right thing but it is worded awkwardly